paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/partition/AbstractPartitionMarkDoneTrigger.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.flink.sink.partition;
+
+import org.apache.paimon.CoreOptions;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.partition.PartitionTimeExtractor;
+import org.apache.paimon.utils.StringUtils;
+
+import org.apache.flink.api.common.state.OperatorStateStore;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_IDLE_TIME_TO_DONE;
+import static org.apache.paimon.flink.FlinkConnectorOptions.PARTITION_TIME_INTERVAL;
+
+/** A Abstract class of PartitionMarkDoneTrigger. */
+public abstract class AbstractPartitionMarkDoneTrigger {
+
+    protected final Map<String, Long> pendingPartitions;
+
+    public AbstractPartitionMarkDoneTrigger(Map<String, Long> pendingPartitions) {
+        this.pendingPartitions = pendingPartitions;
+    }
+
+    public void notifyPartition(String partition) {
+        notifyPartition(partition, System.currentTimeMillis());
+    }
+
+    public void notifyPartition(String partition, long currentTimeMillis) {
+        if (!StringUtils.isNullOrWhitespaceOnly(partition)) {
+            this.pendingPartitions.put(partition, currentTimeMillis);
+        }
+    }
+
+    public List<String> donePartitions() {
```

Comment:
You can introduce a `List<String> donePartitions(boolean endInput)`.

Reply:
done